File: core/src/test/java/dev/morphia/critter/parser/TypesTest.java
package dev.morphia.critter.parser;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Date;
import java.util.Locale;
import java.util.UUID;
import java.util.stream.Stream;

import dev.morphia.critter.parser.gizmo.GizmoExtensions;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.objectweb.asm.Type;

public class TypesTest {

    static Stream<Arguments> types() {
        return Stream.of(
                // Primitive types
                Arguments.of(boolean.class),
                Arguments.of(char.class),
                Arguments.of(byte.class),
                Arguments.of(short.class),
                Arguments.of(int.class),
                Arguments.of(float.class),
                Arguments.of(long.class),
                Arguments.of(double.class),

                // Object types
                Arguments.of(String.class),
                Arguments.of(Locale.class),
                Arguments.of(Date.class),
                Arguments.of(UUID.class),
                Arguments.of(BigDecimal.class),
                Arguments.of(Instant.class),

                // Arrays of boxed primitives (Kotlin Array<X> = Java boxed X[])
                Arguments.of(Boolean[].class),
                Arguments.of(Character[].class),
                Arguments.of(Byte[].class),
                Arguments.of(Short[].class),
                Arguments.of(Integer[].class),
                Arguments.of(Float[].class),
                Arguments.of(Long[].class),
                Arguments.of(Double[].class),

                // Primitive arrays (Kotlin XArray = Java primitive x[])
                Arguments.of(boolean[].class),
                Arguments.of(char[].class),
                Arguments.of(byte[].class),
                Arguments.of(short[].class),
                Arguments.of(int[].class),
                Arguments.of(float[].class),
                Arguments.of(long[].class),
                Arguments.of(double[].class),

                // 2D primitive arrays (Kotlin Array<XArray> = Java primitive x[][])
                Arguments.of(boolean[][].class),
                Arguments.of(int[][].class),

                // Arrays of objects
                Arguments.of(String[].class),
                Arguments.of(Locale[].class),
                Arguments.of(Date[].class),
                Arguments.of(UUID[].class),
                Arguments.of(BigDecimal[].class),
                Arguments.of(Instant[].class),

                // Arrays of arrays (2D)
                Arguments.of(Boolean[][].class),
                Arguments.of(Integer[][].class),
                Arguments.of(String[][].class),
                Arguments.of(Locale[][].class),

                // Arrays of arrays of arrays (3D)
                Arguments.of(Integer[][][].class),
                Arguments.of(String[][][].class));
    }

    @ParameterizedTest
    @MethodSource("types")
    public void asClassConversion(Class<?> expected) {
        Type type = Type.getType(expected);
        Class<?> actual = GizmoExtensions.asClass(type, Thread.currentThread().getContextClassLoader());
        Assertions.assertEquals(expected, actual, "Type " + type.getDescriptor() + " should convert to " + expected.getName());
    }
}

File: critter/core/src/test/kotlin/dev/morphia/critter/parser/TypesTest.kt
package dev.morphia.critter.parser

import dev.morphia.critter.parser.Generators.asClass
import java.math.BigDecimal
import java.time.Instant
import java.util.Date
import java.util.Locale
import java.util.UUID
import org.objectweb.asm.Type
import org.testng.Assert.assertEquals
import org.testng.annotations.DataProvider
import org.testng.annotations.Test

class TypesTest {
    @DataProvider(name = "types")
    fun typeProvider(): Array<Array<Any>> {
        return arrayOf(
            // Primitive types
            arrayOf(Boolean::class.java),
            arrayOf(Char::class.java),
            arrayOf(Byte::class.java),
            arrayOf(Short::class.java),
            arrayOf(Int::class.java),
            arrayOf(Float::class.java),
            arrayOf(Long::class.java),
            arrayOf(Double::class.java),

            // Object types
            arrayOf(String::class.java),
            arrayOf(Locale::class.java),
            arrayOf(Date::class.java),
            arrayOf(UUID::class.java),
            arrayOf(BigDecimal::class.java),
            arrayOf(Instant::class.java),

            // Arrays of boxed primitives
            arrayOf(Array<Boolean>::class.java),
            arrayOf(Array<Char>::class.java),
            arrayOf(Array<Byte>::class.java),
            arrayOf(Array<Short>::class.java),
            arrayOf(Array<Int>::class.java),
            arrayOf(Array<Float>::class.java),
            arrayOf(Array<Long>::class.java),
            arrayOf(Array<Double>::class.java),

            // Primitive arrays
            arrayOf(BooleanArray::class.java),
            arrayOf(CharArray::class.java),
            arrayOf(ByteArray::class.java),
            arrayOf(ShortArray::class.java),
            arrayOf(IntArray::class.java),
            arrayOf(FloatArray::class.java),
            arrayOf(LongArray::class.java),
            arrayOf(DoubleArray::class.java),

            // 2D primitive arrays
            arrayOf(Array<BooleanArray>::class.java),
            arrayOf(Array<IntArray>::class.java),

            // Arrays of objects
            arrayOf(Array<String>::class.java),
            arrayOf(Array<Locale>::class.java),
            arrayOf(Array<Date>::class.java),
            arrayOf(Array<UUID>::class.java),
            arrayOf(Array<BigDecimal>::class.java),
            arrayOf(Array<Instant>::class.java),

            // Arrays of arrays (2D)
            arrayOf(Array<Array<Boolean>>::class.java),
            arrayOf(Array<Array<Int>>::class.java),
            arrayOf(Array<Array<String>>::class.java),
            arrayOf(Array<Array<Locale>>::class.java),

            // Arrays of arrays of arrays (3D)
            arrayOf(Array<Array<Array<Int>>>::class.java),
            arrayOf(Array<Array<Array<String>>>::class.java),
        )
    }

    @Test(dataProvider = "types")
    fun asClassConversion(expected: Class<*>) {
        val type = Type.getType(expected)
        val actual = type.asClass()

        assertEquals(actual, expected, "Type ${type.descriptor} should convert to ${expected.name}")
    }
}
